src/transport/quic.rs

```diff
@@ -0,0 +1,272 @@
+use std::net::SocketAddr;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::Poll;
+use std::time::Duration;
+
+use super::Transport;
+use crate::config::{TlsConfig, TransportConfig};
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use futures_util::{ready, StreamExt};
+use openssl::pkcs12::Pkcs12;
+use quinn::{Connecting, Connection, Endpoint, EndpointConfig, Incoming, NewConnection};
+use rustls::internal::msgs::codec::Codec;
+use rustls::ClientConfig;
+use tokio::fs;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::net::{ToSocketAddrs, UdpSocket};
+use tokio_native_tls::native_tls::Certificate;
+
+pub const ALPN_QUIC_TUNNEL: &[&[u8]] = &[b"qt"];
+pub const NONE: &str = "None";
+
+pub struct QuicTransport {
+    config: TlsConfig,
+    client_crypto: Option<rustls::ClientConfig>,
+}
+
+impl Debug for QuicTransport {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let client_crypto = &self.client_crypto.as_ref().map(|_| "ClientConfig{}");
+
+        f.debug_struct("QuicTransport")
+            .field("config", &self.config)
+            .field("client_crypto", client_crypto)
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct QuicBiStream {
+    send: quinn::SendStream,
+    recv: quinn::RecvStream,
+    conn: Connection,
+}
+
+impl QuicBiStream {
+    fn new((mut send, recv): (quinn::SendStream, quinn::RecvStream), conn: Connection) -> Self {
+        Self { send, recv, conn}
+    }
+}
+
+impl tokio::io::AsyncRead for QuicBiStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        Pin::new(self.get_mut().recv.borrow_mut()).poll_read(cx, buf)
+    }
+}
+
+impl AsyncWrite for QuicBiStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::result::Result<usize, Error>> {
+        Pin::new(self.get_mut().send.borrow_mut()).poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(self.get_mut().send.borrow_mut()).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(self.get_mut().send.borrow_mut()).poll_shutdown(cx)
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<std::result::Result<usize, Error>> {
+        Pin::new(self.get_mut().send.borrow_mut()).poll_write_vectored(cx, bufs)
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        self.send.is_write_vectored()
+    }
+}
+
+pub struct QuicAcceptor(Arc<Mutex<(Endpoint, Incoming)>>);
```

Comment:
If it is possible to change `Transport::accept` to `a: &mut Self::Acceptor`, the mutex overhead should be avoided here.

Reply:
Agreed, would that be ok with @rapiz1 ?

Reply:
I'm ok with that